contracts/BuyBackBurnerBase.sol

```diff
@@ -0,0 +1,140 @@
+
+    struct SingleSwap {
+        bytes32 poolId;
+        SwapKind kind;
+        address assetIn;
+        address assetOut;
+        uint256 amount;
+        bytes userData;
+    }
+
+    struct FundManagement {
+        address sender;
+        bool fromInternalBalance;
+        address payable recipient;
+        bool toInternalBalance;
+    }
+
+    /// @dev Swaps tokens on Balancer.
+    function swap(SingleSwap memory singleSwap, FundManagement memory funds, uint256 limit, uint256 deadline)
+    external payable returns (uint256);
+}
+
+// Bridge interface
+interface IBridge {
+    /// @dev Initiates a withdrawal from L2 to L1 to a target account on L1.
+    /// @param l2Token Address of the L2 token to withdraw.
+    /// @param to Recipient account on L1.
+    /// @param amount Amount of the L2 token to withdraw.
+    /// @param minGasLimit Minimum gas limit to use for the transaction.
+    /// @param extraData Extra data attached to the withdrawal.
+    function withdrawTo(address l2Token, address to, uint256 amount, uint32 minGasLimit, bytes calldata extraData) external;
+}
+
+// ERC20 interface
+interface IERC20 {
+    /// @dev Sets `amount` as the allowance of `spender` over the caller's tokens.
+    /// @param spender Account address that will be able to transfer tokens on behalf of the caller.
+    /// @param amount Token amount.
+    /// @return True if the function execution is successful.
+    function approve(address spender, uint256 amount) external returns (bool);
+
+    /// @dev Gets the amount of tokens owned by a specified account.
+    /// @param account Account address.
+    /// @return Amount of tokens owned.
+    function balanceOf(address account) external view returns (uint256);
+}
+
+interface IOracle {
+    /// @dev Validates price according to slippage.
+    function validatePrice(uint256 slippage) external view returns (bool);
+}
+
+/// @title BuyBackBurnerBase - BuyBackBurner implementation contract for Base
+contract BuyBackBurnerBase is BuyBackBurner {
+    // Token transfer gas limit for L1
+    // This is safe as the value is practically bigger than observed ones on numerous chains
+    uint32 public constant TOKEN_GAS_LIMIT = 300_000;
+
+    // OLAS token address
+    address public olas;
+    // Native token (ERC-20) address
+    address public nativeToken;
+    // Oracle address
+    address public oracle;
+    // L2 token relayer bridge address
+    address public l2TokenRelayer;
+    // Oracle max slippage for ERC-20 native token <=> OLAS
+    uint256 public maxSlippage;
+    // Balancer vault address
+    address public balancerVault;
+    // Balancer pool Id
+    bytes32 public balancerPoolId;
+
+    /// @dev Bridges OLAS amount back to L1 and burns.
+    /// @param olasAmount OLAS amount.
+    /// @param tokenGasLimit Token gas limit for bridging OLAS to L1.
+    /// @return leftovers msg.value leftovers if partially utilized by the bridge.
+    function _bridgeAndBurn(
+        uint256 olasAmount,
```

Comment:
Maybe here instead of passing olasAmount we are going to get the full `olas.balanceOf(address(this))`?

Reply:
Yes imo that's sensible. We should also have a governance controlled lower bound of OLAS that needs to be bridged. It's no good to have tiny amounts of OLAS be bridged individually.

bridge and burn should be logically separate from the swap transaction. Otherwise again, swap requires not too large values, bridge and burn ideally uses very large values to safe on L1 gas